alf/metrics/metrics.py

```diff
@@ -0,0 +1,231 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""A set of metrics.
+Converted to PyTorch from the TF version.
+https://github.com/tensorflow/agents/blob/master/tf_agents/metrics/tf_metrics.py
+"""
+
+from alf.metrics import metric
+import torch
+
+
+class THDeque(torch.nn.Module):
+    """Deque backed by torch.tensor storage."""
+
+    def __init__(self, max_len, dtype):
+        super().__init__()
+        shape = (max_len, )
+        self._dtype = dtype
+        self.max_len = torch.tensor(max_len, dtype=torch.int64)
+        self.buffer = torch.zeros(shape, dtype=dtype)
+        self.head = torch.zeros((), dtype=torch.int64)
+
+    @property
+    def data(self):
+        return self.buffer[:self.length]
+
+    def append(self, value):
+        self.add(value)
+
+    def add(self, value):
+        position = torch.remainder(self.head, self.max_len)
+        self.buffer[position] = value
+        self.head.add_(1)
+
+    @property
+    def length(self):
+        return torch.min(self.head, self.max_len)
+
+    def clear(self):
+        self.head.fill_(0)
+        self.buffer.fill_(0)
+
+    def mean(self):
+        if self.head == 0:
+            return torch.zeros((), dtype=self._dtype)
+        return torch.mean(self.buffer[:self.length])
```

Comment:
I cannot find where self.length is defined/updated. And if it is indeed a bug, unittest is not functional.

Reply:
`length` is a function decorated with `@property`.  Tests did fail when code wasn't correct.  So this should be Ok.